core/src/contracts/stablex_contract.rs

```diff
@@ -377,6 +367,95 @@ async fn get_block(
         .ok_or_else(|| anyhow!("block {:?} is missing", block_number))
 }
 
+trait BatchIdGetter {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>>;
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>>;
+}
+
+impl BatchIdGetter for Web3<DynTransport> {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>> {
+        async move {
+            let current_block = get_block(&self, block_number).await?;
+            Ok(get_block_batch_id(&current_block))
+        }
+        .boxed()
+    }
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>> {
+        async move {
+            let current_block = get_block(&self, BlockNumber::Latest).await?;
+            let batch_id = get_block_batch_id(&current_block);
+            let block_number = current_block
+                .number
+                .ok_or_else(|| {
+                    anyhow!("latest block {:?} has no block number", current_block.hash)
+                })?
+                .as_u64();
+            Ok((batch_id, block_number))
+        }
+        .boxed()
+    }
+}
+
+trait IncrementalBinarySearch<T: BatchIdGetter = Self> {
```

Comment:
```suggestion
trait IncrementalBinarySearch {
```

Reply:
This does not seem needed.

Reply:
Is this trait the meaning behind the acronym "BIS" I see in the PR title?

Reply:
So, from what I understand from this, you want to provide a `get_last_block_for_batch` implementation for any type that implements `BatchIdGetter`. You have two options for this:
```rust
trait BatchIdGetter {
    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>>;
    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>>;

    fn get_last_block_for_batch<'a>(&'a self, batch_id: u32) -> BoxFuture<'a, Result<u64>> {
        // NOTE: Default trait method implementation, types that implement `BatchIdGetter` get this
        // implementation automatically.
        // ...
    }
}
```

Alternatively, if you want to separate it into a separate trait, you could:
```rust
trait IncrementalBinarySearch: BatchIdGetter {
    fn get_last_block_for_batch<'a>(&'a self, batch_id: u32) -> BoxFuture<'a, Result<u64>> {
        // NOTE: Default trait method implementation, types that implement `BatchIdGetter` get this
        // implementation automatically.
        // ...
    }
}
```

Or if you really want to keep it a separate trait, you can do it as @e00E suggested.